### PR TITLE
Support message coalescing

### PR DESCRIPTION
**Pull request by bdraco**
`yarn link` in this folder and then `yarn link home-assistant-js-websocket`

### CODE REVIEW

**Comment by bdraco** [comment]
That yarn add didn't work.  Not sure how to actually test this

**Comment by balloob** [comment]
`yarn link` in this folder and then `yarn link home-assistant-js-websocket`

File: lib/connection.ts [hunk 1]
```diff
@@ -349,57 +349,63 @@ export class Connection {
   }
 
   private _handleMessage = (event: MessageEvent) => {
-    const message: WebSocketResponse = JSON.parse(event.data);
+    let message_group: WebSocketResponse | [WebSocketResponse] = JSON.parse(event.data);
```
Comment (balloob) [inline lib/connection.ts:352]: Current type is list with 1 response in it. This indicates an array:

```suggestion
    let message_group: WebSocketResponse | WebSocketResponse[] = JSON.parse(event.data);
```

**Comment by balloob** [comment]
I think this change is great! I wonder how many race conditions we're going to discover.

File: lib/socket.ts [hunk 2]
```diff
@@ -95,34 +111,38 @@ export function createSocket(options: ConnectionOptions): Promise<HaWebSocket> {
     };
 
     const handleMessage = async (event: MessageEvent) => {
-      const message = JSON.parse(event.data);
-
-      if (DEBUG) {
-        console.log("[Auth phase] Received", message);
+      let message_group: WebSocketAuthMessage | [WebSocketAuthMessage] = JSON.parse(event.data);
```
Comment (balloob) [inline lib/socket.ts:114]: Let's not support message coalescing during auth. I think that it's also a different piece of code in `core` anyway than the one you touched.

**Comment by bdraco** [comment]
I've got to be doing something dumb but I can't figure this one out

```
bdraco@Js-MacBook-Pro home-assistant-js-websocket % cd ~/frontend 
bdraco@Js-MacBook-Pro frontend % yarn link home-assistant-js-websocket
Usage Error: Invalid destination; Can't link the project to itself

$ yarn link [-A,--all] [-p,--private] [-r,--relative] <destination>
bdraco@Js-MacBook-Pro frontend % 
```

**Comment by balloob** [comment]
Hm, you can just do this I guess:

```
rm -rf node_moduels/home-assistant-js-websocket
ln -sf ../home-assistant-js-websocket node_moduels/home-assistant-js-websocket
```

And then it's basically an editable install. Restart webpack.

**Comment by bdraco** [comment]
```
rm -rf node_modules/home-assistant-js-websocket
ln -sf ../home-assistant-js-websocket node_modules/home-assistant-js-websocket
```

Seems to have worked.

Now I think I have a good test setup 👍 

Thanks

**Comment by bdraco** [comment]
Well spoke too soon
```
ERROR in ./src/state/disconnect-toast-mixin.ts 3:0-95
Module not found: Error: Can't resolve 'home-assistant-js-websocket' in '/Users/bdraco/frontend/src/state'
```

**Comment by bdraco** [comment]
I'm going to do `npm build` and copy in the files manually..

**Comment by bdraco** [comment]
that works.

**Comment by bdraco** [comment]
everything works and the latency problem from the connection overhead just disappeared with this change

**Comment by bdraco** [comment]
Testing https://github.com/bdraco/scaletest and I get one message for the event storms it creates 

Makes a big difference on performance

**Comment by bdraco** [comment]
I changed scaletest to do 1000 sensors instead and it no longer overwhelms the system

**Comment by bdraco** [comment]
And my browser is sitting at 1% cpu instead of maxing out

**Comment by bdraco** [comment]
Still high cpu with firefox, but much better

**Review by balloob — APPROVED** [review]

**Comment by balloob** [comment]
We need to add an opt-in message somewhere. Maybe as part of the auth phase ?

**Comment by balloob** [comment]
Ah snap, we also need it to be backwards compatible. So can't do auth phase.

**Comment by bdraco** [comment]
```diff
diff --git a/lib/messages.ts b/lib/messages.ts
index 57258c1..fd655e4 100644
--- a/lib/messages.ts
+++ b/lib/messages.ts
@@ -4,6 +4,7 @@ export function auth(accessToken: string) {
   return {
     type: "auth",
     access_token: accessToken,
+    features: ["coalesce_messages"],
   };
 }
 
```

I was thinking something like this but it blows up auth

**Comment by bdraco** [comment]
Sending `supported_features` right after auth worked.

Kinda meh on the naming though

Screen Shot 2022-08-24 at 3 15 59 PM
Screen Shot 2022-08-24 at 3 16 05 PM

File: lib/messages.ts [hunk 3]
```diff
@@ -7,6 +7,14 @@ export function auth(accessToken: string) {
   };
 }
 
+export function supported_features() {
+  return {
+    type: "supported_features",
+    id: 1, // Always the first message after auth
+    features: { coalesce_messages: 1.0 },
```
Comment (bdraco) [inline lib/messages.ts:14]: I versioned the features in case we ever have `coalesce_messages` 2.0 etc

**Comment by bdraco** [comment]
For testing I disabled the dev repo in `frontend:` to make sure it still works

before

**Comment by bdraco** [comment]
Here is the after
after

**Comment by balloob** [comment]
We should limit sending supported features based on the HA version that we get back. We can access the version string on the socket. https://github.com/home-assistant/home-assistant-js-websocket/blob/master/lib/socket.ts#L20

**Comment by balloob** [comment]
Otherwise people get warnings in their logs on older HA versions.

**Comment by bdraco** [comment]
Doesn't it already do that?

**Comment by balloob** [comment]
oops missed that.

File: lib/socket.ts [hunk 4]
```diff
@@ -112,6 +113,10 @@ export function createSocket(options: ConnectionOptions): Promise<HaWebSocket> {
           socket.removeEventListener("close", closeMessage);
           socket.removeEventListener("error", closeMessage);
           socket.haVersion = message.ha_version;
+          if (atLeastHaVersion(socket.haVersion, 2022, 9, 0)) {
```
Comment (balloob) [inline lib/socket.ts:116]: you can leave `patch` undefined. I think that otherwise it will not work in dev. `Number("0.dev0")` results in NaN

Comment (balloob) [reply lib/socket.ts:116]: https://github.com/home-assistant/home-assistant-js-websocket/blob/master/lib/util.ts#L59

```suggestion
          if (atLeastHaVersion(socket.haVersion, 2022, 9)) {
```

**Review by balloob — APPROVED** [review]

File: lib/messages.ts (hunk 3 above)
Comment (bdraco) [inline lib/messages.ts:14]:
```suggestion
    features: { coalesce_messages: 1 },
```

**Comment by bdraco** [comment]
Retesting looks good 👍

**Comment by zachowj** [comment]
Not sure if there is a standard enforcement, but you're using snake case where the existing code base is using camel case.

**Comment by bdraco** [comment]
> 



> Not sure if there is a standard enforcement, but you're using snake case where the existing code base is using camel case.

That's what happens when you right more python 😄 

Adjusted

**Pull request merged**